<a href="https://colab.research.google.com/github/Aidin12/wiki-scraper-json-extractor/blob/main/wikipedia_scrapes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
!pip install selenium
!pip install Scrapy
!apt update
!apt install chromium-chromedriver
!cp /usr/lib/chromium-browser/chromedriver /usr/bin

     |████████████████████████████████| 911kB 5.7MB/s 
     |████████████████████████████████| 245kB 4.1MB/s 
     |████████████████████████████████| 256kB 7.5MB/s 
     |████████████████████████████████| 61kB 4.4MB/s 
     |████████████████████████████████| 3.2MB 8.0MB/s 
     |████████████████████████████████| 3.2MB 39.1MB/s 
     |████████████████████████████████| 3.1MB 43.1MB/s 
     |████████████████████████████████| 81kB 5.9MB/s 
  Created wheel for PyDispatcher: filename=PyDispatcher-2.0.5-cp37-none-any.whl size=11517 sha256=c9f5f09e17b72bff072a4ba535a72901a9927b205edb5a71f48461f8041a8fd2
  Stored in directory: /root/.cache/pip/wheels/88/99/96/cfef6665f9cb1522ee6757ae5955feedf2fe25f1737f91fa7f
  Created wheel for protego: filename=Protego-0.1.16-cp37-none-any.whl size=7766 sha256=8199d044fd3827d53f2e8e26bfe475ed7dbde7fbd7052b687b5f37f11c0d215c
  Stored in directory: /root/.cache/pip/wheels/51/01/d1/4a2286a976dccd025ba679acacfe37320540df0f2283ecab12
Successfully built PyDispatche

In [ ]:
# %cd gdrive/My Drive/

In [ ]:
%cd /content/gdrive/My Drive/Wikipedia/

/content/gdrive/My Drive/Wikipedia


In [ ]:
from scrapy import Spider
from selenium import webdriver
from scrapy.selector import Selector #to gather urls from entire site
from scrapy.http import Request
from time import sleep
from selenium.common.exceptions import NoSuchElementException
from selenium import webdriver
options = webdriver.ChromeOptions()
options.add_argument('--headless')
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')

ModuleNotFoundError: ignored

In [ ]:
# !scrapy startproject wiki_crawler

In [ ]:
# !cd wiki_crawler

In [ ]:
# !scrapy genspider wikispider pageviews.toolforge.org

In [ ]:
import scrapy
from scrapy.crawler import CrawlerProcess
from scrapy.linkextractors import LinkExtractor
import pandas as pd

!gdown --id 1M98aMkqwssJtjXUZFKgJlbNU6FN2hoH0

sp500 = pd.read_csv('SP500.csv')
sp500_tickers = list(sp500.Ticker)

ModuleNotFoundError: ignored

In [ ]:
from datetime import datetime

class wikiTrends():

  def __init__(self, start, end, search):
    self.search = search
    self.st = datetime.strptime('01/01/16', '%m/%d/%y')
    self.en = datetime.strptime('01/03/21', '%m/%d/%y')
    self.smonth = self._padZero(self.st.month)
    self.emonth = self._padZero(self.en.month)
    self.sday = self._padZero(self.st.day)
    self.eday = self._padZero(self.en.day)

  def getURL(self):
    s = f'{self.st.year}-{self.smonth}-{self.sday}'
    e = f'{self.en.year}-{self.emonth}-{self.eday}'
    #terms = f'pages={self.search}_(NASDAQ)|{self.search}(NASDAQ)|{self.search}'
    terms = f'pages={self.search}'

    url_start = 'https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start='
    full_url = f'{url_start}{s}&end={e}&{terms}'
    return full_url

  def _padZero(self, date):
      if len(str(date)) == 1:
          return '0'+str(date)

In [ ]:
class WikiSpider(Spider):

    #print(url)
    name = 'wiki'
    allowed_domains = ['pageviews.toolforge.org']
    #start_urls = url

    def __init__(self, url=None, *args, **kwargs):
        super(WikiSpider, self).__init__(*args, **kwargs)
        self.start_urls = url

    def start_requests(self):

        self.driver = webdriver.Chrome(options=options)
        self.driver.get(self.start_urls)

        try:
            download1 = self.driver.find_element_by_xpath('//html/body/main/div[2]/div/div')
            download2 = self.driver.find_element_by_xpath('//html/body/main/div[2]/div/div/ul/li[1]/a')
            sleep(1)
            self.logger.info('downloading')
            download1.click()
            sleep(1)
            download2.click()
            sleep(1)

        except NoSuchElementException:
            self.driver.quit()
            print("Faulty URL")

    def parse_wiki(self, response):
        pass


In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')
%cd /content/gdrive/My Drive/Wikipedia/

import glob
from scrapy import Spider
from selenium import webdriver
from scrapy.selector import Selector #to gather urls from entire site
from scrapy.http import Request
from time import sleep
from selenium.common.exceptions import NoSuchElementException
from selenium import webdriver
options = webdriver.ChromeOptions()
options.add_argument('--headless')
options.add_argument('--no-sandbox')
options.add_argument('--disable-dev-shm-usage')

import scrapy
from scrapy.crawler import CrawlerProcess
from scrapy.linkextractors import LinkExtractor
import pandas as pd

!gdown --id 1M98aMkqwssJtjXUZFKgJlbNU6FN2hoH0

sp500 = pd.read_csv('SP500.csv')
sp500_tickers = list(sp500.Ticker)

from datetime import datetime

class wikiTrends():

  def __init__(self, start, end, search):
    self.search = search
    self.st = datetime.strptime('01/01/16', '%m/%d/%y')
    self.en = datetime.strptime('01/03/21', '%m/%d/%y')
    self.smonth = self._padZero(self.st.month)
    self.emonth = self._padZero(self.en.month)
    self.sday = self._padZero(self.st.day)
    self.eday = self._padZero(self.en.day)

  def getURL(self):
    s = f'{self.st.year}-{self.smonth}-{self.sday}'
    e = f'{self.en.year}-{self.emonth}-{self.eday}'
    terms = f'pages={self.search}_(NASDAQ)|{self.search}(NASDAQ)|{self.search}'
    #terms = f'pages={self.search}'

    url_start = 'https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start='
    full_url = f'{url_start}{s}&end={e}&{terms}'
    return full_url

  def _padZero(self, date):
      if len(str(date)) == 1:
          return '0'+str(date)

class WikiSpider(Spider):

    #print(url)
    name = 'wiki'
    allowed_domains = ['pageviews.toolforge.org']
    #start_urls = url

    def __init__(self, url=None, *args, **kwargs):
        super(WikiSpider, self).__init__(*args, **kwargs)
        self.start_urls = url

    def start_requests(self):

        self.driver = webdriver.Chrome(options=options)
        self.driver.get(self.start_urls)

        try:
            download1 = self.driver.find_element_by_xpath('//html/body/main/div[2]/div/div')
            download2 = self.driver.find_element_by_xpath('//html/body/main/div[2]/div/div/ul/li[1]/a')
            sleep(1)
            self.logger.info('downloading')
            download1.click()
            sleep(1)
            download2.click()
            sleep(1)

        except NoSuchElementException:
            self.driver.quit()
            print("Faulty URL")

    def parse_wiki(self, response):
        pass


Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).
/content/gdrive/My Drive/Wikipedia
Downloading...
From: https://drive.google.com/uc?id=1M98aMkqwssJtjXUZFKgJlbNU6FN2hoH0
To: /content/gdrive/My Drive/Wikipedia/SP500.csv
100% 85.3k/85.3k [00:00<00:00, 24.4MB/s]


In [ ]:
import glob

scraped = glob.glob("*.csv")
scraped

['SP500.csv',
 'AAPL_2016_2021.csv',
 'MSFT_2016_2021.csv',
 'AMZN_2016_2021.csv',
 'FB_2016_2021.csv',
 'TSLA_2016_2021.csv',
 'SNE_2016_2021.csv']

In [ ]:
file_name = 'AAPL_2016_2021.csv'
if file_name in scraped:
  print(True)

True


In [ ]:
sp500_tickers = ['AAPL','MSFT','AMZN','FB','TSLA']
print(len(sp500_tickers))

5


In [ ]:
import os
from itertools import cycle

sp500_tickers = ['AAPL','MSFT','AMZN','FB']
scraped = glob.glob("*.csv")
#print(len(sp500_tickers))

pool = cycle(sp500_tickers)
count = 0

for ticker in pool:
    file_name = f'{ticker}_2016_2021.csv'

    if file_name in scraped:
      count += 1
      print(file_name, count)
      if count == len(sp500_tickers):
        print('Done', count)
        break

    else:
      count = 0

AAPL_2016_2021.csv 1
MSFT_2016_2021.csv 2
AMZN_2016_2021.csv 3
FB_2016_2021.csv 4
Done 4


In [ ]:
print(len(sp500_tickers))

506


In [ ]:
import os
from itertools import cycle

# Set parameters MM/DD/YY
start = '01/01/16'
end = '01/03/21'

#sp500_tickers = ['AAPL','MSFT','AMZN','FB','TSLA','SNE']
scraped = glob.glob("*.csv")
print(len(sp500_tickers))

unable_to_scrape = []

pool = cycle(sp500_tickers)
count = 0

for ticker in pool:
    scraped = glob.glob("*.csv")
    print(f'[INFO] Scraped {len(scraped)} tickers - {scraped}')
    file_name = f'{ticker}_2016_2021.csv'
    print(f'[INFO] Looking for {file_name}')

    if file_name in scraped:
      count += 1
      print(f'[INFO] {file_name} scraped already count = {count}')
      if count == len(sp500_tickers):
        print('Done', count)
        break

    else:
      count = 0
      wk = wikiTrends(start, end, ticker)
      link = wk.getURL()
      print(f'[INFO] Scraping {ticker} - {link}')
      process = CrawlerProcess()
      process.crawl(WikiSpider, url=link)
      process.start()
      sleep(15)

      try:
        os.rename('pageviews-20160101-20210103.csv', file_name)
      except:
        unable_to_scrape.append(ticker)
        print('[INFO] Failed to rename file')
        pass

print(unable_to_scrape)

In [ ]:
print(unable_to_scrape)

NameError: ignored

In [ ]:
import os
from itertools import cycle

# Set parameters MM/DD/YY
start = '01/01/16'
end = '01/03/21'

pool = cycle(sp500_tickers)
count = 0

for ticker in pool:
    file_name = f'{ticker}_2016_2021.csv'

    if file_name not in scraped:
      count = 0
      wk = wikiTrends(start, end, ticker)
      link = wk.getURL()
      print(f'Scraping {ticker} - {link}')
      process = CrawlerProcess()
      process.crawl(WikiSpider, url=link)
      process.start()
      sleep(15)

      try:
        os.rename('pageviews-20160101-20210103.csv', file_name)
      except:
        pass
    else:
      count += 1
      if count == len(sp500_tickers)-1:
        break

2021-03-24 00:27:22 [scrapy.utils.log] INFO: Scrapy 2.4.1 started (bot: scrapybot)
2021-03-24 00:27:22 [scrapy.utils.log] INFO: Versions: lxml 4.2.6.0, libxml2 2.9.8, cssselect 1.1.0, parsel 1.6.0, w3lib 1.22.0, Twisted 21.2.0, Python 3.7.10 (default, Feb 20 2021, 21:17:23) - [GCC 7.5.0], pyOpenSSL 20.0.1 (OpenSSL 1.1.1j  16 Feb 2021), cryptography 3.4.6, Platform Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic
2021-03-24 00:27:22 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.epollreactor.EPollReactor
2021-03-24 00:27:22 [scrapy.crawler] INFO: Overridden settings:
{}
2021-03-24 00:27:22 [scrapy.extensions.telnet] INFO: Telnet Password: 8e1f477049c36cdb
2021-03-24 00:27:22 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.logstats.LogStats']
2021-03-24 00:27:22 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downlo

Scraping AAPL - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=AAPL


2021-03-24 00:27:23 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57251/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:27:23 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:57251
2021-03-24 00:27:23 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57251 "POST /session HTTP/1.1" 200 720
2021-03-24 00:27:23 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:27:23 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57251/session/30f7405b7477d7e05aadb50cc54f25e0/url {"url": "https://pageviews.toolforge.or

Scraping MSFT - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=MSFT


2021-03-24 00:27:44 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:53777/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:27:44 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:53777
2021-03-24 00:27:44 [urllib3.connectionpool] DEBUG: http://127.0.0.1:53777 "POST /session HTTP/1.1" 200 720
2021-03-24 00:27:44 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:27:44 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:53777/session/d494543b16728a670abd1c8f592335b9/url {"url": "https://pageviews.toolforge.or

Scraping AMZN - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=AMZN


2021-03-24 00:28:05 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:53123/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:28:05 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:53123
2021-03-24 00:28:05 [urllib3.connectionpool] DEBUG: http://127.0.0.1:53123 "POST /session HTTP/1.1" 200 720
2021-03-24 00:28:05 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:28:05 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:53123/session/8e29ac4c931aad2c3a01ade0799f9bae/url {"url": "https://pageviews.toolforge.or

Scraping FB - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=FB


2021-03-24 00:28:26 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:45927/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:28:26 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:45927
2021-03-24 00:28:26 [urllib3.connectionpool] DEBUG: http://127.0.0.1:45927 "POST /session HTTP/1.1" 200 720
2021-03-24 00:28:26 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:28:26 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:45927/session/3b66764fbba8f980b1c3d62e516524c6/url {"url": "https://pageviews.toolforge.or

Scraping GOOGL - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=GOOGL


2021-03-24 00:28:47 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:35413/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:28:47 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:35413
2021-03-24 00:28:47 [urllib3.connectionpool] DEBUG: http://127.0.0.1:35413 "POST /session HTTP/1.1" 200 720
2021-03-24 00:28:47 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:28:47 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:35413/session/2e53d66118550afc09a1064f01d77ddb/url {"url": "https://pageviews.toolforge.or

Scraping GOOG - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=GOOG


2021-03-24 00:29:08 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:38879/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:29:08 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:38879
2021-03-24 00:29:09 [urllib3.connectionpool] DEBUG: http://127.0.0.1:38879 "POST /session HTTP/1.1" 200 720
2021-03-24 00:29:09 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:29:09 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:38879/session/470bc4a2cea86eb7ed3e52438f03179d/url {"url": "https://pageviews.toolforge.or

Scraping TSLA - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=TSLA


2021-03-24 00:29:29 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:48167/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:29:29 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:48167
2021-03-24 00:29:30 [urllib3.connectionpool] DEBUG: http://127.0.0.1:48167 "POST /session HTTP/1.1" 200 720
2021-03-24 00:29:30 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:29:30 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:48167/session/2e9b46f6d7a37fb1729b7bae8f066154/url {"url": "https://pageviews.toolforge.or

Scraping BRKB - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=BRKB


2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:
2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:
2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:
2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:
2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:
2021-03-24 00:29:50 [twisted] CRITICAL: Unhandled error in Deferred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/scrapy/crawler.py", line 192, in crawl
    return self._crawl(crawler, *args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/scrapy/crawler.py", line 196, in _crawl
    d = crawler.crawl(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/twisted/internet/defer.py", line 1656, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/usr/local/lib/python3.7/dist-packages/twisted/internet/defer.py", line 1571, in _cancellableInlineCallbacks
    _inlineCallbacks(No

Scraping JPM - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=JPM


2021-03-24 00:30:12 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:55265/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:30:12 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:55265
2021-03-24 00:30:12 [urllib3.connectionpool] DEBUG: http://127.0.0.1:55265 "POST /session HTTP/1.1" 200 720
2021-03-24 00:30:12 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:30:12 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:55265/session/c2d7764256f343796755409afbee9f9e/url {"url": "https://pageviews.toolforge.or

Scraping JNJ - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=JNJ


2021-03-24 00:30:33 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:36081/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:30:33 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:36081
2021-03-24 00:30:33 [urllib3.connectionpool] DEBUG: http://127.0.0.1:36081 "POST /session HTTP/1.1" 200 720
2021-03-24 00:30:33 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:30:33 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:36081/session/37a34d280ac92f2f2f8ae1eae8c5823f/url {"url": "https://pageviews.toolforge.or

Scraping V - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=V


2021-03-24 00:30:54 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:39529/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:30:54 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:39529
2021-03-24 00:30:54 [urllib3.connectionpool] DEBUG: http://127.0.0.1:39529 "POST /session HTTP/1.1" 200 720
2021-03-24 00:30:54 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:30:54 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:39529/session/971c8c84a3d60ba86d2d740e2eb2a8b8/url {"url": "https://pageviews.toolforge.or

Scraping DIS - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=DIS


2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:31:15 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:57541
2021-03-24 00:31:15 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session HTTP/1.1" 200 720
2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session/a23da30722752e672d8841a97e05bf67/url {"url": "https://pageviews.toolforge.or

FileNotFoundError: ignored

Scraping DIS - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=DIS
2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session {"capabilities": {"firstMatch": [{}], "alwaysMatch": {"browserName": "chrome", "platformName": "any", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}, "desiredCapabilities": {"browserName": "chrome", "version": "", "platform": "ANY", "goog:chromeOptions": {"extensions": [], "args": ["--headless", "--no-sandbox", "--disable-dev-shm-usage"]}}}
2021-03-24 00:31:15 [urllib3.connectionpool] DEBUG: Starting new HTTP connection (1): 127.0.0.1:57541
2021-03-24 00:31:15 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session HTTP/1.1" 200 720
2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:31:15 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session/a23da30722752e672d8841a97e05bf67/url {"url": "https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=DIS"}
2021-03-24 00:31:16 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session/a23da30722752e672d8841a97e05bf67/url HTTP/1.1" 200 14
2021-03-24 00:31:16 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:31:16 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session/a23da30722752e672d8841a97e05bf67/element {"using": "xpath", "value": "//html/body/main/div[2]/div/div"}
2021-03-24 00:31:16 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session/a23da30722752e672d8841a97e05bf67/element HTTP/1.1" 200 88
2021-03-24 00:31:16 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:31:16 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session/a23da30722752e672d8841a97e05bf67/element {"using": "xpath", "value": "//html/body/main/div[2]/div/div/ul/li[1]/a"}
2021-03-24 00:31:17 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session/a23da30722752e672d8841a97e05bf67/element HTTP/1.1" 200 88
2021-03-24 00:31:17 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request
2021-03-24 00:31:18 [wiki] INFO: downloading
2021-03-24 00:31:18 [selenium.webdriver.remote.remote_connection] DEBUG: POST http://127.0.0.1:57541/session/a23da30722752e672d8841a97e05bf67/element/f72143ba-75c1-4703-98d7-7564dc0ae840/click {"id": "f72143ba-75c1-4703-98d7-7564dc0ae840"}
2021-03-24 00:31:18 [urllib3.connectionpool] DEBUG: http://127.0.0.1:57541 "POST /session/a23da30722752e672d8841a97e05bf67/element/f72143ba-75c1-4703-98d7-7564dc0ae840/click HTTP/1.1" 400 181
2021-03-24 00:31:18 [selenium.webdriver.remote.remote_connection] DEBUG: Finished Request

In [ ]:
#sp500_tickers = ['AAPL','MSFT','AMZN','FB','SNE']

In [ ]:
wk = wikiTrends(start, end, ticker)
link = wk.getURL()
print(f'Scraping {ticker} - {link}')
process = CrawlerProcess()
process.crawl(WikiSpider, url=link)
process.start()

2021-03-24 01:04:15 [scrapy.utils.log] INFO: Scrapy 2.4.1 started (bot: scrapybot)
2021-03-24 01:04:15 [scrapy.utils.log] INFO: Versions: lxml 4.2.6.0, libxml2 2.9.8, cssselect 1.1.0, parsel 1.6.0, w3lib 1.22.0, Twisted 21.2.0, Python 3.7.10 (default, Feb 20 2021, 21:17:23) - [GCC 7.5.0], pyOpenSSL 20.0.1 (OpenSSL 1.1.1j  16 Feb 2021), cryptography 3.4.6, Platform Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic
2021-03-24 01:04:15 [scrapy.utils.log] DEBUG: Using reactor: twisted.internet.epollreactor.EPollReactor
2021-03-24 01:04:15 [scrapy.crawler] INFO: Overridden settings:
{}
2021-03-24 01:04:15 [scrapy.extensions.telnet] INFO: Telnet Password: 8cea1fe7d37c075a
2021-03-24 01:04:15 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.memusage.MemoryUsage',
 'scrapy.extensions.logstats.LogStats']


Scraping AAPL - https://pageviews.toolforge.org/?project=en.wikipedia.org&platform=all-access&agent=user&redirects=0&start=2016-01-01&end=2021-01-03&pages=AAPL


2021-03-24 01:04:15 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2021-03-24 01:04:15 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.ref